### PR TITLE
Budget App Capstone Rails

### PR DESCRIPTION
**Pull request by Maxwell011**
In this milestone I have :

- Installed Ruby on Rails framework.
- Understood Rails RESTful design and router.
- Used controllers to handle requests and render empty views.
- Used params from browser requests in a safe way.
- Used preprocessed HTML file with embedded Ruby code.
- Used layouts and templates for shared content.
- Used database migration files to maintain database schema.
- Used validations for models.
- Understood what ORM is.
- Written SQL queries with ActiveRecord.
- Set up associations between models
- Built a mobile web app that requires the user to log in.]Used devise gem for authentication.
- Limited access to web app resources based on authorization rules.
- Analyzed in writing why you have made a coding choice using one structure over another.
- Implemented unit and integrations tests for main functionalities

### Required Links
[Loom video](https://www.loom.com/share/db745b1ed9a34aa385144b8e134ebdf5)
[Deployed link](https://budget-app-nm3z.onrender.com)

Kindly Review 🙏

### CODE REVIEW

**Review by EricMbouwe — CHANGES_REQUESTED** [review]
Hi @Maxwell011, 

Good job so far! 🔥
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## Suggested changes ♻
- [ ] There are some failing tests. Let's fix that

Screenshot 2023-03-03 at 11 31 10 PM

_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: spec/models/users_spec.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+RSpec.describe User, type: :model do
+  before :each do
+    @user = User.new(name: 'John Doe', email: 'test@email.com', password: 'secret')
+  end
+
+  context 'Testing Validations' do
+    it 'is valid with valid attributes' do
+      @user.save
+      expect(@user).to_not be_valid
+    end
```
Comment (KDlamini) [inline spec/models/users_spec.rb]: You still have one test that is failing. I do not see any new commits after the previous review that were pushed to consolidate the failing test. 

image

image

- Kindly push your changes.

Problem:

The issue is in the file `./spec/models/users_spec.rb` line 11 where the `@user` is expected not be valid and yet it is valid, that is why the test is failing.

File: spec/requests/categories_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    get groups_path
+  end
+
+  it 'is a success' do
+    expect(response).to have_http_status(:ok)
+  end
+
+  it "renders 'index' template" do
+    expect(response).to render_template('index')
+  end
```
Comment (KDlamini) [inline spec/requests/categories_spec.rb:20]: [OPTIONAL] You can cover more integrations specs here where you can go further and for example:

- Test the visibility of categories on the page
- Test the visibility of transactions in at least one category
- Test the create new category button
- Test the create new transaction button

File: spec/requests/splash_spec.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe 'Splashes', type: :request do
+  describe 'GET /index' do
+    it 'returns http success' do
+      get '/splash/index'
+      expect(response).to have_http_status(:success)
+    end
+  end
+end
```
Comment (KDlamini) [inline spec/requests/splash_spec.rb:10]: [OPTIONAL] Same here, you can include more tests where you use Capybara to simulate the registration and login process as well as include cases where registration and login are invalid.

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi @Maxwell011 🙋‍♀️,  

You have done an incredible job implementing some of the changes requested by the previous reviewer. 👏🏾👏🏾👏🏾  
However, there are still some changes that aren't addressed yet. You are almost there! 💪🏾 

## HIGHLIGHTS

✔️ Well-written model tests
✔️ Good PR title and description

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

N/A

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Maxwell011** [comment]
This Test passes on my end and I don't know why it doesn't on yours
here is it
Screenshot (37)

**Comment by Maxwell011** [comment]
Kindly Reach out to me please

File: spec/models/users_spec.rb (hunk 1 above)
Comment (DeliceLydia) [reply spec/models/users_spec.rb]: - [ ] Please address this as requested with the previous reviewer 
Screen Shot 2023-03-04 at 19 37 39

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
## Hi @Maxwell011, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.


 ## Highlights
- Application is working fine

 ## Required Changes ♻️
_Check the comments under the review._




### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KDlamini — APPROVED** [review]
### Hi @Maxwell011 🙋‍♀️,  

Excellent work! 💯 
You have done a great job implementing all the required changes. 🎊 🥳 
Your feature is complete! There is nothing else to say other than... it's time to merge it. :shipit: 
Congratulations!!!

## HIGHLIGHTS 🌟 

✔️ Great PR title and description
✔️ Models, controllers, routes, and views set up correctly
✔️ User authentic works as expected
✔️ Correct use of GitHub flow

# Status: Approved ✔️ 

oscars-standing-ovation
 
Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**